Adds the slippery component

SteelSlayer opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
Loosely ports TG's slippery component, but its been heavily modified for paradise, and has some of my own additions as well, namely creating a base `after_slip()` proc for *all* movable atoms.

There is no CL because this is a backend change only.
<!-- Include a small to medium description of what your PR changes. Document every changes or this may delay review or even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Cuts down on `obj` vars related to slipping that go unused for a majority of items. More OOP.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

AffectedArc07 commented on code/datums/components/slippery.dm:
```diff
@@ -0,0 +1,54 @@
+/**
+	This is a component that can be applied to any movable atom (mob or obj).
+
+	While the atom has this component, any carbon mob that walks over it will slip.
+	Duration, tiles moved, and so on, depend on what variables are passed in when the component is added.
+*/
+/datum/component/slippery
+	/// Text that gets displayed in the slip proc, i.e. "user slips on [description]"
+	var/description
+	/// The amount of stun to apply after slip.
+	var/stun
+	/// The amount of weaken to apply after slip.
+	var/weaken
+	/// The chance that walking over the parent will slip you.
+	var/slip_chance
+	/// The amount of tiles someone will be moved after slip.
+	var/slip_tiles
+	/// TRUE If this slip can be avoided by walking.
+	var/walking_is_safe
+	/// TRUE if having no slip shoes makes you immune to this slip.
+	var/noslip_is_immune
+	/// The verb that players will see when someone slips on the parent. In the form of "You [slip_verb]ped on".
+	var/slip_verb
```
The documentation is great for the var definitions, but not for the definition.

While I would love to suggest reading up on the doucmentation guide, its not compulsory here so its no huge issue

SteelSlayer replied:
I can make this *not* autodoc'd if you want, as in `/* bla */`

AffectedArc07 replied:
```
/**
	This is a component that can be applied to any movable atom (mob or obj).
	While the atom has this component, any carbon mob that walks over it will slip.
	Duration, tiles moved, and so on, depend on what variables are passed in when the component is added.
*/
```

Should be
```
/**
  * # Slip Component
  *
  * This is a component that can be applied to any movable atom (mob or obj).
  *
  * While the atom has this component, any carbon mob that walks over it will slip.
  * Duration, tiles moved, and so on, depend on what variables are passed in when the component is added.
  *
  */
```

but again, doc isnt required here so its up to you

SteelSlayer replied:
Eh I really hate that asterisk per line format but, If that's how its suppose to be, why not

Fox-McCloud commented:
`but its been heavily modified for paradise, `

Whenever you do a port, but heavily modify it, you should enumerated differences.

Some of us have a high familiarity with TG code and what it does, so we expect functionality to be the same. It's a pain in the ass when we have to manually find the differences ourselves.

Fox-McCloud commented on code/datums/components/slippery.dm:
```diff
@@ -0,0 +1,57 @@
+	var/slip_verb
+
+/datum/component/slippery/Initialize(_description, _stun = 0, _weaken = 0, _slip_chance = 100, _slip_tiles = 0, \
+									_walking_is_safe = TRUE, _noslip_is_immune = TRUE, _slip_verb = "slip")
```
Just put this all on teh same line, please.

SteelSlayer replied:
Sure

Fox-McCloud commented on code/datums/components/slippery.dm:
```diff
@@ -0,0 +1,57 @@
+
+/datum/component/slippery/Initialize(_description, _stun = 0, _weaken = 0, _slip_chance = 100, _slip_tiles = 0, \
+									_walking_is_safe = TRUE, _noslip_is_immune = TRUE, _slip_verb = "slip")
+	if(!ismovable(parent))
```
`ismovableatom`

We'll have to wait until 513 stable and use updating to that, officially before we can do this.

SteelSlayer replied:
Got it.

SteelSlayer replied:
I realized this should be `isatom` because I think this component is made to be placed on turfs as well.

SteelSlayer commented:
Sorry, I can list out the differences from TG. 
- All of the component's variables are different so they correspond with the variables that are removed from `item`'s and the arguments for the `slip` proc. Such as the slip chance and amount of tiles moved.
- This: `RegisterSignal(parent, COMSIG_ITEM_WEARERCROSSED, .proc/Slip_on_wearer)` and the proc in there have been removed, because we don't have this signal.
- I removed the `callback` var because there didn't seem to be a use for it. However, I was exploring around TG the other day and I saw that its used for `wet_floor` component so I may need to add that back in, I don't know.
- The creation of the `atom/movable/after_slip()` proc was my idea.
- TG has a `!victim.is_flying()` check among the other checks on my line 55. But I removed it since we check that in the `slip` proc anyway, and `slip` is used outside this component.

Fox-McCloud commented on code/game/atoms_movable.dm:
```diff
@@ -330,6 +330,16 @@
+
+	If you want the person who slipped to have something special done to them, put it here.
+*/
+/atom/movable/proc/after_slip(mob/living/carbon/human/H)
```
if the component is atom level, this should be as well.

Fox-McCloud commented:
Last thing. 

Our components aren't 100% up to date with TG's, so we can't make use of named args.

https://github.com/tgstation/tgstation/pull/49098

Make sure that everything works without utilizing named args.

SteelSlayer commented:
I doubled checked and did a final test, all the args should be in right spot for all the items.
